targets/TARGET_Ambiq_Micro/TARGET_Apollo3/device/pinmap.c

```diff
@@ -0,0 +1,117 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mbed_assert.h"
+#include "pinmap.h"
+#include "mbed_error.h"
+
+#include "extensions.h"
+#include "am_mcu_apollo.h"
+
+void pin_config(PinName pin, am_hal_gpio_pincfg_t pincfg)
+{
+    if (pin == (PinName)NC) {
+        return;
+    }
+    MBED_ASSERT(AM_HAL_STATUS_SUCCESS == am_hal_gpio_pinconfig(pin, pincfg));
+}
+
+void pinmap_config(PinName pin, const PinMap *map)
+{
+    // fully configure a pin by a pin map entry
+    if (pin == NC) {
+        return;
+    }
+
+    am_hal_gpio_pincfg_t pincfg;
+    while (map->pin != NC) {
+        if (map->pin == pin) {
+            pincfg = *((am_hal_gpio_pincfg_t *)(map->function));
+            pin_config(pin, pincfg);
+            return;
+        }
+        map++;
+    }
+    MBED_ERROR1(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_CODE_PINMAP_INVALID), "could not pinmap_config", pin);
+}
+
+void pin_function(PinName pin, int function)
+{
+    // am_hal_gpio_pincfg_t cfg = {0};
+    // cfg.uFuncSel = function;
+    // am_hal_gpio_pinconfig((uint32_t)(pin), cfg); // apply configuration
+
+#define PADREG_FLD_FNSEL_S 3
+
+    uint32_t ui32Padreg;
+    uint32_t ui32Funcsel;
+
+    MBED_ASSERT(pin < AM_HAL_GPIO_MAX_PADS);
+
+    //
+    // Initialize the PADREG accumulator variables.
+    //
+    ui32Padreg = 0;
+
+    // Get the requested function
+    ui32Funcsel = (uint32_t)function;
+
+    ui32Padreg |= ui32Funcsel << PADREG_FLD_FNSEL_S;
+
+    //
+    // At this point, the configuration variable ui32Padreg,
+    //  values is set (at bit position 0) and ready to write
+    //  to its respective register bitfields.
+    //
+    uint32_t ui32PadregAddr;
+    uint32_t ui32PadClearMask;
+    uint32_t ui32PadShft;
+
+    ui32PadregAddr = AM_REGADDR(GPIO, PADREGA) + (pin & ~0x3);
+
+    ui32PadShft = ((pin & 0x3) << 3);
+    ui32PadClearMask = ~((uint32_t)0x38 << ui32PadShft);
+
+    // Get the new values into their rightful bit positions.
+    ui32Padreg <<= ui32PadShft;
+
+    AM_CRITICAL_BEGIN
+
+    GPIO->PADKEY = GPIO_PADKEY_PADKEY_Key;
+
+    AM_REGVAL(ui32PadregAddr) = (AM_REGVAL(ui32PadregAddr) & ui32PadClearMask) | ui32Padreg;
+
+    GPIO->PADKEY = 0;
+
+    AM_CRITICAL_END
+}
+
+void pin_mode(PinName pin, PinMode mode)
+{
+    MBED_ASSERT(0);
+    // gpio_t obj = {
```

Comment:
why is this commented out, shall we remove it? 
what should pin mode do ? being empty? It might be then worth adding a comment why is this function empty

Reply:
Hey, I am trying to remember the full story here. If I remember right, a lot of targets try to use the function pinmap_pinout when initializing different drivers. This calls pin_function, and pin_mode. At first we tried to implement this way, but ended up ditching it to match closer with the ambiq hal functionality, which should do both of these things in a single function. I think we had problems with how the settings from these 2 things interacted with eachother, and wanted the error checking and validation provided by the ambiq functions we ended up utilizing. We left the MBED_ASSERT(0) in this function so that we would know if our approach had caused any problems and left the dead code as a starting point should we turn back. I've removed the dead code and left a comment in its place explaining what is above.

Reply:
I didn't like this answer after providing it. I feel like we are required to provide this function even if we don't use it ourselves. I have created an implementation of this function and pushed it up.